### PR TITLE
Add JPA based StackDao implementation

### PR DESCRIPTION
**Pull request by voievodin**
### What does this PR do?

Adds a new StackDao implementation based on JPA.
Adds tests for StackDao interface contract(TCKs).
Exposes TCKs and uses it for current local(json based) implementation.
### What issues does this PR fix or reference?
#1795
### Tests written?

Yes
### Docs requirements?

No

@skabashnyuk, @akorneta, @sleshchenko please review

### CODE REVIEW

File: core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/repository/JpaTckRepository.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Override
+    public void removeAll() throws TckRepositoryException {
+        uow.begin();
+        final EntityManager manager = managerProvider.get();
+        manager.getTransaction().begin();
+        try {
+            // The query 'DELETE FROM Entity' won't be correct as it will ignore orphanRemoval
+            // and may also ignore some configuration options, while EntityManager#remove won't
+            manager.createQuery(format("SELECT e FROM %s e", getEntityName(entityClass)), entityClass)
```
Comment (skabashnyuk) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/repository/JpaTckRepository.java:80]: what about DELETE \* FROM  ?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaStackDao.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                                  .setFirstResult(skipCount)
+                                  .getResultList();
+        } catch (RuntimeException x) {
+            throw new ServerException(x.getLocalizedMessage(), x);
+        }
+    }
+
+    @Transactional
+    protected void doCreate(StackImpl stack) {
+        managerProvider.get().persist(stack);
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaStackDao.java:120]: why you need to have do\* methods?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/StackImpl.java [hunk 3]
```diff
@@ -20,36 +20,93 @@
...
- * Server implementation of {@link Stack}
+ * Data object for {@link Stack}.
  *
  * @author Alexander Andrienko
+ * @author Yevhenii Voevodin
  */
+@Entity(name = "Stack")
+@NamedQueries(
+        @NamedQuery(name = "Stack.search",
+                    query = "SELECT DISTINCT(stack) " +
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/StackImpl.java]: why distinct?
Can you give me an analysis tree for postgresql?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaStackDao.java (hunk 2 above)
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaStackDao.java:120]: `@Transactional` intercepts action with something like 

```
tx.begin();
action();
tx.commit();
```

from the context `tx.commit()` throws `DuplicateKeyException` and `RollbackException`hierarchies which we need to catch and rethrow as `ConflcitException` and so on. If i put `@Transactional` directly on method it wouldn't be possible to catch errors.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/StackImpl.java (hunk 3 above)
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/StackImpl.java]: `distinct` is odd, thank you

File: core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/repository/JpaTckRepository.java (hunk 1 above)
Comment (voievodin) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/repository/JpaTckRepository.java:80]: Read the comment above the code

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/StackImpl.java (hunk 3 above)
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/StackImpl.java]: h2 `EXPLAIN`

```
SELECT
    T0.ID AS A1,
    T0.CREATOR AS A2,
    T0.DESCRIPTION AS A3,
    T0.NAME AS A4,
    T0.SCOPE AS A5,
    T0.ORIGIN AS A6,
    T0.TYPE AS A7,
    T0.DATA AS A8,
    T0.MEDIATYPE AS A9,
    T0.ICON_NAME AS A10,
    T0.WORKSPACECONFIG_ID AS A11
FROM PUBLIC.STACK T0
    /* PUBLIC.STACK.tableScan */
WHERE 'search' IN(
    SELECT DISTINCT
        T2.PUBLICACTIONS
    FROM PUBLIC.STACK_PUBLICACTIONS T2
        /* PUBLIC.FK_STACK_PUBLICACTIONS_STACK_ID_INDEX_3: STACK_ID = T0.ID */
        /* WHERE T0.ID = T2.STACK_ID
        */
    INNER JOIN PUBLIC.STACK T1
        /* PUBLIC.PRIMARY_KEY_4: ID = T0.ID
            AND ID = T2.STACK_ID
         */
        ON 1=1
    WHERE ((T0.ID = T1.ID)
        AND (T2.STACK_ID = T1.ID))
        AND (T0.ID = T2.STACK_ID))
```

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaStackDao.java (hunk 2 above)
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaStackDao.java:120]: ok

**Comment by skabashnyuk** [comment]
Can you give me example how Stack.search query is looks like for postgresql and postgresql EXPLAIN output.

File: core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/repository/JpaTckRepository.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            return clazz.getSimpleName();
+        }
+        final Entity entity = clazz.getAnnotation(Entity.class);
+        if (entity.name().isEmpty()) {
+            return clazz.getSimpleName();
+        }
+        return entity.name();
+    }
+}
+
```
Comment (sleshchenko) [inline core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/tck/repository/JpaTckRepository.java]: remove extra line

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/stack/StackService.java [hunk 5]
```diff
@@ -260,7 +260,7 @@ public Response uploadIcon(@ApiParam("The image for stack")
                                @ApiParam("The stack id")
                                @PathParam("id")
                                final String id)
-            throws NotFoundException, ServerException, BadRequestException, ForbiddenException {
+            throws NotFoundException, ServerException, BadRequestException, ForbiddenException, ConflictException {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/stack/StackService.java:263]: Add description of ConflictException to swagger doc

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java [hunk 6]
```diff
@@ -11,25 +11,74 @@
...
+
+        bind(new TypeLiteral<TckRepository<WorkspaceImpl>>() {}).to(WorkspaceTckRepository.class);
+        bind(new TypeLiteral<TckRepository<StackImpl>>() {}).to(StackTckRepository.class);
+
+        bind(WorkspaceDao.class).to(JpaWorkspaceDao.class);
+        bind(StackDao.class).to(JpaStackDao.class);
+    }
+
+    public static class WorkspaceTckRepository extends JpaTckRepository<WorkspaceImpl> {
+        @Inject
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java]: Why do we need @Inject here?

**Comment by sleshchenko** [comment]
LGTM

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java (hunk 6 above)
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java]: `JpaTckRepository` injects `EntityManager` and `UnitOfWork`, `JpaTckRepository` is an abstract class, so lets assume that `@Inject` is removed, how to bind `WorkspaceTckRepository` ? `bind(...).toInstance(new WorkspaceTckRepository())` won't work, as it won't inject fields to the superclass,  on the other hand i think it is not really good to use `requestInjection` for this purpose, so i prefer `@Inject` to `requestInjection`

Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java]: Thanks for explanation.
FYI I made default constructor as public and it worked for me without Inject annotation with binding to class `bind(new TypeLiteral<TckRepository<WorkspaceImpl>>() {}).to(WorkspaceTckRepository.class);`

Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java]: Thank you, i will rework it in proposed way

**Comment by voievodin** [comment]
> Can you give me example how Stack.search query is looks like for postgresql and postgresql EXPLAIN output.

```
EXPLAIN SELECT t0.ID, t0.CREATOR, t0.DESCRIPTION, t0.NAME, t0.SCOPE, t0.ORIGIN, t0.TYPE, t0.DATA, t0.MEDIATYPE, t0.icon_name, t0.WORKSPACECONFIG_ID FROM STACK t0 WHERE ((2 = 0) OR (2 = (SELECT COUNT(t2.TAGS) FROM Stack_TAGS t2, STACK t1 WHERE (((t2.TAGS IN ('search-tag1', 'search-tag2')) AND (t0.ID = t1.ID)) AND (t2.Stack_ID = t1.ID)))))

"Seq Scan on stack t0  (cost=0.00..202.38 rows=1 width=4684) (actual time=0.092..0.213 rows=2 loops=1)"
"  Filter: (2 = (SubPlan 1))"
"  Rows Removed by Filter: 3"
"  SubPlan 1"
"    ->  Aggregate  (cost=19.22..19.23 rows=1 width=516) (actual time=0.037..0.037 rows=1 loops=5)"
"          ->  Nested Loop  (cost=0.14..19.21 rows=1 width=516) (actual time=0.025..0.033 rows=1 loops=5)"
"                ->  Seq Scan on stack_tags t2  (cost=0.00..11.05 rows=1 width=1032) (actual time=0.015..0.018 rows=1 loops=5)"
"                      Filter: (((tags)::text = ANY ('{search-tag1,search-tag2}'::text[])) AND ((stack_id)::text = (t0.id)::text))"
"                      Rows Removed by Filter: 18"
"                ->  Index Only Scan using stack_pkey on stack t1  (cost=0.14..8.15 rows=1 width=516) (actual time=0.008..0.008 rows=1 loops=6)"
"                      Index Cond: (id = (t0.id)::text)"
"                      Heap Fetches: 6"
"Planning time: 0.347 ms"
"Execution time: 0.297 ms"

```

@skabashnyuk

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java [hunk 7]
```diff
@@ -11,25 +11,67 @@
...
+                    manager.persist(new UserImpl(id, id + "email", id + "-name"));
+                }
+                manager.persist(entity);
+            }
+        }
+
+        @Override
+        public void removeAll() throws TckRepositoryException {
+            final EntityManager manager = managerProvider.get();
+            manager.createQuery("SELECT stack FROM Stack stack", StackImpl.class).getResultList().forEach(manager::remove);
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java]: can you use named "Stack.getAll" query here?

Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/jpa/WorkspaceTckModule.java]: yes

**Comment by skabashnyuk** [comment]
ok

**Pull request merged**